File: java/src/main/java/com/mongodb/tradecraft/sequence_server/SequenceServer.java
package com.mongodb.tradecraft.sequence_server;

import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import static com.mongodb.client.model.Filters.eq;
import com.mongodb.client.model.FindOneAndUpdateOptions;
import com.mongodb.client.model.ReturnDocument;
import static com.mongodb.client.model.Updates.inc;
import org.bson.Document;
import spark.Request;
import spark.Response;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class SequenceServer {

    Logger logger;

    private MongoClient mongoClient;
    private MongoDatabase database;
    private MongoCollection<Document> collection;
    private static int x=-1;

    SequenceServer() {
        mongoClient = MongoClients.create("mongodb+srv://smartek:smartekpwd@smartek.2ec6p.mongodb.net/sequencedb?retryWrites=true");
        database = mongoClient.getDatabase("sequencedb");
        collection = database.getCollection("sequenceCol");
        System.out.println("Database contains " + collection.countDocuments() + " sequences");
        logger = LoggerFactory.getLogger(SequenceServer.class);
    }

    //Fetch back inforamtion about samples
    public String sequence(Request req, Response res) {
        res.type("application/json");
        String name = req.params(":name");
         x=(x+1)%10;
         name=name.concat("_"+x);
//        Integer counter = updateDocumentAndgetCounter(name);
        
        String json = "{" + name + ":" + updateDocumentAndgetCounter(name)+ "}";
        res.type("application/json");
        return json;
    }

    public Integer updateDocumentAndgetCounter(String name) {
        FindOneAndUpdateOptions options = new FindOneAndUpdateOptions();
        options.upsert(true);
        options.returnDocument(ReturnDocument.AFTER);
       
        Document doc = collection.findOneAndUpdate(eq("_id",name ), inc("counter", 1), options);
        return doc.getInteger("counter")*10+x;
    }

}

File: node/sequenceserver.js
const express = require('express');
const { Db } = require('mongodb');
const MongoClient = require("mongodb").MongoClient;
const app = express()

const port = 6789
const connString= "mongodb+srv://smartek:smartekpwd@smartek.2ec6p.mongodb.net/sequencedb?retryWrites=true"
const client= new MongoClient(connString)

let db
let collection


client.connect(function(err){
    if(err==null){
         db=client.db("sequencedb")
         collection=db.collection("sequenceCol")
    //    client.close
    }
})


function sequence(req,res) {
    name = req.params['name']
    
    collection.findOneAndUpdate({"_id":name},{"$inc":{"counter":1}},{upsert:true,returnOriginal:false},function(err,result){
        if(err==null){
            let response={}
            // counter=result["value"]
            counter=result.value["counter"]
            response[name]=counter
            res.send(response)
            // return result

        }
    })
    
   
    
    //  let doc= await collection.fi
 
}

app.put('/sequence/:name', sequence)

app.listen(port, () => console.log(`Example app listening on port ${port}!`))



        